### PR TITLE
feat: add hierarchical namespace and folders features

### PR DESCRIPTION
**Pull request by JesseLovelace**
Adds support for hierarchical namespace and folders. Copied over from an internal workspace

### CODE REVIEW

**Review by BenWhitehead — COMMENTED** [review]
Superficially LGTM, waiting on @frankyn for final approval as I believe he's more familiar with this

**Review by frankyn — APPROVED** [review]

**Pull request merged**
